src/dists.ad.js

```diff
@@ -1347,14 +1351,90 @@ var Marginal = makeDistributionType({
     return this.supp;
   },
   print: function() {
-    return _.map(this.params.dist, function(obj, val) { return [val, obj.prob]; })
-        .sort(function(a, b) { return b[1] - a[1]; })
-        .map(function(pair) { return '    ' + pair[0] + ' : ' + pair[1]; })
-        .join('\n');
+    return printDist(this.params.dist);
   }
 });
 
 
+// A "list of samples" backed marginal distribution that only
+// aggregates the samples into a distribution when necessary.
+var SampleBasedMarginal = makeDistributionType({
+  name: 'Marginal',
+  nodoc: true,
+  nohelper: true,
+  params: [{name: 'samples'}],
+  mixins: [finiteSupport],
+  constructor: function() {
+    if (!_.isArray(this.params.samples) ||
+        this.params.samples.length === 0) {
+      throw new Error('Expected samples to be a non-empty array.');
+    }
+
+    // Provide access to the samples using the interface previously
+    // provided by the `justSample` option.
+    // samples is an array of objects like: {value: ..., score: ...}
+    this.samples = this.params.samples;
+
+    this.getDist = function() {
+      if (this._cacheddist) {
+        return this._cacheddist;
+      } else {
+        var dist = {};
+        this.params.samples.forEach(function(obj) {
+          var val = obj.value;
+          var k = util.serialize(val);
```

Comment:
Use a different letter? I don't think there's much danger of confusion here, but since `k` is almost exclusively used for continuations, maybe other uses are worth avoiding.